### ISSUE TITLE
`no-extra-semi` mistakenly removes necessary semicolon when `semi` is `never` 

### ISSUE THREAD

**Issue by flowmemo** [eslint/eslint#7928]
**Tell us about your environment**

* **ESLint Version:**
3.13.1
* **Node Version:**
6.9.3
* **npm Version:**
4.1.1
**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**

```yaml
rules:
  semi:
    - error
    - never
  no-extra-semi:
    - error
```

**What did you do? Please include the actual source code causing the issue.**

```js
var foo = function(){};
;[1].map(foo)
```

**What did you expect to happen?**
```js
var foo = function(){}
;[1].map(foo)
```
**What actually happened? Please include the actual, raw output from ESLint.**
```js
var foo = function(){}
[1].map(foo)
```

When `semi` is `never`, `no-extra-semi` is true, and there are whitespaces between two semicolons, those rules will delete both of semicolons. `semi` deletes the first one, and `no-extra-semi` deletes the second one.

**Comment by not-an-aardvark** [comment]
Thanks for the report -- I was able to reproduce this issue.

I'm not sure what the correct fix for this should be. Two problems are getting autofixed here (`semi` and `no-extra-semi` both remove extra semicolons), and both fixes are independently correct. However, the behavior of the code changes when both fixes are applied simultaneously. We might need to make the conflict detection more strict in [`source-code-fixer`](https://github.com/eslint/eslint/blob/7e04b33028e68555c69cf2568d606cdeab0b7cda/lib/util/source-code-fixer.js).

**Comment by mysticatea** [comment]
Thank you for the report.

> We might need to make the conflict detection more strict in source-code-fixer.

I agree.
Or, maybe we can revisit #3885.

**Comment by alberto** [comment]
> Two problems are getting autofixed here (semi and no-extra-semi both remove extra semicolons), and both fixes are independently correct. However, the behavior of the code changes when both fixes are applied simultaneously. 

I think we should remove the overlapping behaviour from `semi`.

**Comment by not-an-aardvark** [comment]
@alberto I disagree. The rules are both working correctly, it's just that their fixes happen to interfere with each other.  `semi` removes the first semicolon (which is correct behavior, because the first semicolon is unnecessary in the original source text). `no-extra-semi` removes the second semicolon (which is also correct behavior, because the second semicolon is also unnecessary in the original source text). Neither rule knows anything about what the other rule is doing, so core is the only good place to handle this.

If we start adding exceptions to specific fixers to prevent them from conflicting with fixers from other rules, then we're going to end up with a lot of duplicated logic. For example, the `semi` rule will have to detect `no-extra-semi` errors in order to determine whether it should handle the fix. With hundreds of rules, this will quickly become very difficult to reason about, and it also won't work for custom autofixable rules.

Core is designed to prevent this by isolating fixes. If a rule makes a fix that is safe on its own, core is supposed to guarantee that the fix will be safe when applied with other fixes. This bug violates the guarantee, so it's an issue with core, not with the rules.

**Comment by alberto** [comment]
Sorry, after re-reading the original issue I think I misinterpreted it. I'm still not sure it's an autofixing issue, rather than a reporting issue in the first place (I think there is just one problem in the context of the program, not two), but I'm not sure how I would prevent that. Except by merging the rules, as @mysticatea mentioned.

Anyhow, I think one approach to avoid the conflicting fixes would be to consider the range including both semicolons as part of the fix in `no-extra-semi`.

Do you have something in mind regarding `source-code-fixer`?

**Comment by not-an-aardvark** [comment]
> Do you have something in mind regarding `source-code-fixer`?

I think we could fix issues like this by making conflict detection more strict. We currently detect conflicts between fixes based on whether their ranges overlap -- maybe we could change that to detect conflicts if there are any tokens between the ranges.

**Comment by alangpierce** [comment]
This also came up in #8026, where no-else-return and no-useless-return have independently-correct fixes that together break the code.

I'd be happy to help contribute a fix to this if there's agreement on the approach (or at least happy to brainstorm solutions).

Reading through the code a bit, here's my understanding of how all this stuff works (please correct me if I'm missing/misunderstanding anything):

* Each rule optionally advertises one fix of the form (start, end, text), meaning that the range from start to end should be replaced with the text.
* The code in source-code-fixer.js applies as many of these as it can, working backwards, and ignores any fixes that happen on overlapping ranges.
* The `multipassFix` function in cli-engine.js re-runs eslint and all fixes over and over, up to 10 times, until no more fixing can be done. This generally means that any fixes will make it through eventually, even if there are some conflicts.

My impression is that we need to extend the autofixer API in some way, since it seems impossible for source-code-fixer to have enough context as-is to know if two changes are incompatible.

Here's one idea: in addition to returning the start, end, and text, any rule can optionally specify a range of code that must not be touched in the same pass.
* no-extra-semi and semi could maybe specify the entire range between the preceding and following statements (so that we don't remove two semicolons in a row).
* no-useless-return and maybe no-else-return could specify the entire function body, since any change to the control flow could cause correctness issues.

That would force the autofixes to be applied independently, which I think fixes both of these cases.

I guess another approach might be to just make the replacement bigger for these rules, e.g. no-useless-return would replace the entire function body with the same function body minus the `return`. That would fix things without needing any changes to source-code-fixer and without needing any changes to the fix format. I guess a downside to this is that it seems pretty subtle and non-obvious, so it would at least need better documentation for rule-writers, and maybe an extension to the rule-fixer API to make it easier.

**Comment by ilyavolodin** [comment]
@alangpierce Adding additional locking information to the fix will help in some cases (like the one we are talking about here), but it will cause some valid, non-conflicting autofixes to be skipped. In your example, if we lock the whole range between preceding and following statements `linebreak-style` rule will not be able to modify windows line ending to unix format, since line break operator will be between two statements. In general, there's a good chance that you will prevent whitespace rules to modify the code with this approach.
I guess maybe it's better to not fix some issues, but guarantee correct output? I'm not 100% sure.

P.S. I'm most definitely don't want to discourage you from brainstorming solutions. Just pointing out issues that might come up from this approach.

**Comment by not-an-aardvark** [comment]
@alangpierce Your analysis of how autofixing works is basically correct. One additional note: fixable rules also must export a `meta.fixable` property, which has a value of either `code` or `whitespace`. (At the moment, the property isn't actually used for anything, but it was introduced in case we ever needed to filter fixes to avoid conflicts, e.g. by only applying fixes that don't affect the AST.)

Your idea of specifying a "don't touch this code" range for fixes seems like it would work. My main concern is that it might be difficult for a rule author to think of all the possible conflicting fixes that could cause a problem. I like that the current autofixing system doesn't have much mental overhead -- a rule can (supposedly) just convert code to other valid code, and assume that core will take care of any conflicts.

---

Here are a few other possibilities (some of which might have their own issues, but listed for the sake of brainstorming):

1. Automatically infer the conflict range based on the range of the reported node. (This is similar to your suggestion of using larger fix ranges. I think your suggestion is probably better, because this would tightly couple the report range with the internal fix logic.)
1. Infer the conflict range based on something else from a rule, such as the nodes that the rule listens for. (Since rules can use `parent` properties or `sourceCode.ast`, this probably wouldn't be reliable.)
1. Give up on trying to detect conflicts, and reparse after applying every fix. (This has all the desirable qualities that we would want, except that the performance would probably be unacceptable.)
1. Include metadata about fixes and the AST when detecting conflicts. For example, I would assume we can be more liberal about fixes that don't modify the AST, or fixes that only modify whitespace. (If we can figure out a good way to do this, I think it would be the best solution.)

**Comment by alangpierce** [comment]
@ilyavolodin I think both fixes would still happen in your example (if I'm understanding things right). All that the locking would ensure is that the two fixes aren't applied *in the same pass*. In this case, the source-code-fixer code would pick either `linebreak-style` or `semi` to apply in the first pass, and skip the other, and in the second pass, the other fix would be detected and applied.

This might make eslint slower on average (e.g. 3 or 4 passes instead of 2) and more likely to hit the 10 pass limit, but my impression is that it won't be a really big deal, and that correctness for autofixing is really important.

**Comment by alangpierce** [comment]
@not-an-aardvark 

> My main concern is that it might be difficult for a rule author to think of all the possible conflicting fixes that could cause a problem.

Agreed, although I think that autofixing does fundamentally require a lot of thought, at least in cases that modify the AST. (For example, I know `no-var` has accumulated a lot of special cases to deal with various correctness issues.)

One way of thinking about my "don't touch this code" range idea is that it's a compromise between your option #3 and the current (incorrect) system. If every rule chose the whole program as the "don't touch this code" range, then the rules would be applied sequentially, but rules can specify a more specific range to say "I'm only affecting this part of the code". This allows the autofixing process to parallelize some things while running rules sequentially when possible. With that mindset, the best advice I would have to rule authors is to be conservative about what range to specify.

While we're brainstorming, some other possible approaches:

5. The fix specifies an enum-like value with one of a few situations, e.g. "this won't conflict with other fixes", "this might conflict with other fixes in the enclosing function", and "this might conflict with other fixes in the program".

6. The fix optionally specifies a node that is the "domain" of the fix, and the "don't touch this code" range is the range of that node.

These both would make things a little more straightforward for rule authors, I think. But neither of those would work really nicely with the semicolon conflict for this bug.

**Comment by not-an-aardvark** [comment]
As far as I can tell, there are two distinct types of fix conflicts here:

1. Fixes between adjacent tokens that don't affect the AST (e.g. the conflict between `semi` and `no-extra-semi` in this issue)
1. Fixes that change the AST based on the presence of code that gets changed by another fix (e.g. https://github.com/eslint/eslint/issues/8026)

If I'm not mistaken, the first type of conflict should be possible to detect correctly since JS grammar is context-free. I think we can assume there is no conflict between non-AST fixes if there is at least one token between the fix ranges.

Detecting AST-based conflicts is harder, and it might not be solvable in the general case without duplicating rule logic (unless the rule provides explicit ranges). That said, I think the majority of fixable rules don't touch the AST. From a performance perspective, it might be feasible to do the following:

1. Parse the source text and get a list of fixes
1. If there are any fixes that would modify the AST, apply the *first* such fix. Otherwise, apply all fixes that don't affect the AST (with the conflict detection scheme above).
1. Repeat until no fixable problems are reported

**Comment by alberto** [comment]
IIRC we apply whitespace fixes first, so that wouldn't really be a problem.

**Comment by mysticatea** [comment]
:thinking: This seems a big issue.

Now, I'm remembering #7348.
I guess we may solve this problem if fixers can return multiple fix-units. Assuming it to allow fixes which do no change (it replace a range with the original text).

- `semi (never)` returns the fix of target semicolons and the next token of every target semicolon (if the next token is a semicolon). The next token's fixes would prevent the removing of the next token.
- `no-else-return` returns the fix of target `else` blocks and the ReturnStatements of sibling `if`/`else` blocks. The ReturnStatement's fixes would prevent the change of code paths.

**Comment by alangpierce** [comment]
Cool discussion. @not-an-aardvark's breakdown in https://github.com/eslint/eslint/issues/7928#issuecomment-277595227 makes sense to me. Just to clarify something that confused me a little, "fixes that don't affect the AST" really means "fixes that wouldn't affect the AST if applied in isolation". Removing a double-semicolon like in this bug certainly does affect the AST. But AST-preserving fixes that are far enough apart won't collectively change the AST due to JS being context-free. It's worth noting that I think this case still fits pretty well into my "don't touch this code" range idea: each AST-preserving fix just "locks" the code range of its surrounding tokens.

> If there are any fixes that would modify the AST, apply the first such fix. Otherwise, apply all fixes that don't affect the AST (with the conflict detection scheme above).

I'm worried that serializing all modifications that change the AST would make perf way worse at least in some cases. In the most extreme cases for me, I might have an autogenerated (translated from CoffeeScript) file that's thousands of line long with hundreds of `prefer-const` violations. (That's the most common AST-changing autofix that I can think of, although there may be others.) But also, perf is less of a big deal in my case because I'm working on a one-time code transform rather than something you'd run regularly while developing.

I guess another open question is how to know whether a fixer affects the AST. Maybe it could be another `meta.fixable` value? That certainly seems more straightforward than trying to detect it automatically.

Given that there are a lot of open questions here and that it's possible to solve these bugs without any API changes, my impression is that it's best to do that for now and see how things play out. Like I mentioned at the bottom of https://github.com/eslint/eslint/issues/7928#issuecomment-277560675, both of these bugs can be fixed by specifying a larger code range that mostly stays the same (similar to how the `no-extra-parens` fixer replaces all code between the parens, even though it only removes the parens). The existing conflict detection code would then effectively treat that code range as being "locked" for that pass. And maybe in the future the API could be extended to make it a little easier for rule authors to avoid these types of conflicts.

(Regarding multiple fix units (#7348), I'm less convinced that that's worth it. The only reason to do so is to avoid conflicts for perf reasons, since any multi-fix can always be expressed as one large replacement, and it seems like it adds a lot of complexity. Having a cleaner API for rule authors would of course be nice, though. For example, a rule might say "remove this open paren and that close paren" and that would end up as a single replacement replacing everything in between.)

**Comment by not-an-aardvark** [comment]
> Just to clarify something that confused me a little, "fixes that don't affect the AST" really means "fixes that wouldn't affect the AST if applied in isolation". Removing a double-semicolon like in this bug certainly does affect the AST.

Correct, I meant fixes that don't affect the AST in isolation. Sorry about the confusion.

> I'm worried that serializing all modifications that change the AST would make perf way worse at least in some cases. In the most extreme cases for me, I might have an autogenerated (translated from CoffeeScript) file that's thousands of line long with hundreds of `prefer-const` violations.

Good point, I'd forgotten that rules like `prefer-const` can suggest multiple fixes at a time. It's probably reasonable to require rules to ensure that their *own* fixes don't conflict with each other (i.e. if a rule suggests two fixes, the rule should make sure it's possible for both of them to be applied). So we could get around this by grouping all the fixes for a single rule, which would allow all the `prefer-const` autofixes get applied in a single pass.

> I guess another open question is how to know whether a fixer affects the AST. Maybe it could be another `meta.fixable` value? That certainly seems more straightforward than trying to detect it automatically.

Yes, that was what I was thinking as well. We could deprecate/repurpose the existing `code` and `whitespace` options, and have something like this (bikesheddable):

```js
module.exports = {
  meta: {
    fixable: {
      modifiesAST: true
    }
  },
  create(context) {
    // ...
  }
};
```

**Comment by not-an-aardvark** [comment]
cc'ing @nzakas on this discusison since he implemented the current conflict-detection code.

**Comment by alangpierce** [comment]
> It's probably reasonable to require rules to ensure that their *own* fixes don't conflict with each other

👍 great point. I expect perf would be much less of a worry if we grouped all fixes for the same rule.

**Comment by alangpierce** [comment]
I think I should be able to come up with a PR to fix both of these bugs using my strategy in the last paragraph of https://github.com/eslint/eslint/issues/7928#issuecomment-277560675 , unless there are objections. That is, I'll extend the replacement range in the fixes for `no-extra-semi`, `semi`, `no-else-return`, and `no-useless-return` so that the conflict detection process applies the conflicting fixes in different passes. `no-extra-semi` and `semi` will "lock" the code range from the previous to the next token, and `no-else-return` and `no-useless-return` will "lock" their enclosing function.

At the very least, you could call it a stopgap solution, since it fixes both known bugs without requiring any API changes. But I think I'm fairly happy with the "don't touch this code" idea, at least for advanced use cases.

One thought is that a future API change might allow a rule to a specify a "don't touch this code" range, but it would be optional. For AST-modifying changes, if that range is unspecified, it defaults to "locking" the entire program. That way, the default behavior for a new rule is that it is run in isolation, and advanced rule authors and popular rules (like the built-in ones) can specify more detail so that they can be safely run in parallel with other rules.

**Comment by not-an-aardvark** [comment]
@alangpierce For the time being, I think expanding the range for those rules is a good workaround. Feel free to make a PR.

I am still a bit concerned about having a "don't touch this code" (DNTC) range -- in addition to the my concerns about increasing mental overhead, I'm worried that a rule author who didn't quite understand the API might end up setting the range too low, causing the rule's fixer to break code. However, I like your idea of defaulting to locking the entire program -- that way core is safe by default (similar to the idea in https://github.com/eslint/eslint/issues/7928#issuecomment-277595227), and users who know what they're doing can specify a DNTC range for a performance boost. 

I'm also interested in hearing feedback from others on this approach.

**Comment by nzakas** [comment]
A few random thoughts:

1. We are now stretching the autofix mechanism further than it was originally intended to go, and we are seeing problems related to it. It's probably a good reminder that we should be thinking about how to better do autofixes in the future. 
2. I don't see performance as a big concern so long as we only add 1-2 new fix application cycles. People definitely are willing to trade a reasonable amount of time for autofixing. Plus, the biggest hit is running autofixing for the first time, subsequent runs likely have much less to fix. And then files are probably autofixing one or a few at a time. 
3. I'm a bit concerned about a "do not touch" system, as that seems ripe for abuse by people either misunderstanding the purpose or intentionally being disruptive to the ecosystem. It also requires rule authors to have an understanding of how their rule interacts with other existing rules, which I think is a significant burden that is unfair to force upon them. 
4. I think trying to say if a rule modifies the AST is too complex a solution due to the intricacies of parsing. 
5. Expanding the range of the possible fix seems like a good approach to test out before publishing for custom rule authors to use. My only concern here is that this range info is contingent on knowledge about how some other rule works, and again, that's a lot for rule authors to have to have in their heads to use this feature correctly. I'd suggest trying this out in a PR to see just how much you're relying on understanding how another rule is working. Can you rationalize the change you're making without referencing how another rule works? If not, we probably need to think through the problem a bit more. 

One of the key aspects of ESLint rules is that they have no knowledge of what other rules exist or are doing. Any encroachment on that, no matter how small, starts us down a path of interdependence that will make rules tougher to deal with, so let's keep that in mind.

**Comment by platinumazure** [comment]
Just out of curiosity, how often do these fix conflicts occur when the actual fixes aren't right next to each other?

**Comment by not-an-aardvark** [comment]
@nzakas, thanks for your thoughts. Regarding a few of the things you mentioned:

> I don't see performance as a big concern so long as we only add 1-2 new fix application cycles. People definitely are willing to trade a reasonable amount of time for autofixing.

I agree that performance is not a big concern for autofixing. I'll note that if we do decide to go forward with the "modifies AST" heuristic from https://github.com/eslint/eslint/issues/7928#issuecomment-277595227, it might be significantly more than 1-2 new fix cycles. In total, we would end up with one autofix pass for each AST-modifying rule that reports an error, followed by a couple more passes for formatting fixes to be applied.

> [A "do not touch" system] requires rule authors to have an understanding of how their rule interacts with other existing rules, which I think is a significant burden that is unfair to force upon them.

I mostly agree with this. Given that these conflicts existed for awhile before we noticed them, it seems unrealistic to expect rules to be able to anticipate conflicts on their own.

> I think trying to say if a rule modifies the AST is too complex a solution due to the intricacies of parsing.

I'm confused about why this would be complex. To clarify, rules would be indicating this explicitly in metadata. It seems like it should be simple for rule authors to determine whether a rule affects the AST, since rule authors understand the semantics of the fixes they produce. 

> Expanding the range of the possible fix seems like a good approach to test out before publishing for custom rule authors to use. My only concern here is that this range info is contingent on knowledge about how some other rule works, and again, that's a lot for rule authors to have to have in their heads to use this feature correctly. I'd suggest trying this out in a PR to see just how much you're relying on understanding how another rule is working. Can you rationalize the change you're making without referencing how another rule works? If not, we probably need to think through the problem a bit more.

I share your concern here. Actually, https://github.com/eslint/eslint/pull/8067 is a PR that does this -- it sets the range of fixes for a few rules to cover the entire enclosing function. This does fix the particular problems that we know about with core rules, but it probably won't work for all cases.

---

Generally, I want to point out that a rule might be able to specify a "do not touch" range without needing to consider interactions with other rules. Conflicts arise when a rule uses a piece of code to determine whether it should report an error, but that piece of code isn't included in the fix range. If another rule modifies that piece of code, the fix might become wrong. So a rule could reasonably indicate a range of code that it took into account when deciding to report an error, and that doesn't require knowledge or consideration of other rules.

**Comment by not-an-aardvark** [comment]
@platinumazure 

> Just out of curiosity, how often do these fix conflicts occur when the actual fixes aren't right next to each other?

They appear to be somewhat rare -- the conflict between `no-else-return` and `no-useless-return` is the only one we know about between core rules.

**Comment by platinumazure** [comment]
@not-an-aardvark Thanks. That's unfortunate, I was hoping a lookahead/lookbehind approach could be used (e.g., semi would not return a fix descriptor if the next token was an open bracket/parentheses), but that won't be sufficient.

**Comment by mysticatea** [comment]
> Generally, I want to point out that a rule might be able to specify a "do not touch" range without needing to consider interactions with other rules. Conflicts arise when a rule uses a piece of code to determine whether it should report an error, but that piece of code isn't included in the fix range. If another rule modifies that piece of code, the fix might become wrong. So a rule could reasonably indicate a range of code that it took into account when deciding to report an error, and that doesn't require knowledge or consideration of other rules.

I agree with this. For examples:

- `no-else-return` relies on the code path in the previous `if` block. So it can get safe if it locks control flow statements in the `if` block.
- `no-useless-return` relies on the fact that any statement does not exist on the code path preceded by the `return` statement. So it can get safe if it locks the range between the `return` statement and the end of the containing function.
- `no-extra-semi` relies on the fact the semi is an empty statement. It can get safe if it locks between the last token of the previous statement and the semi.

However, there are rules which rely on a wide range, so it may become "1 pass per a report" in fact. In that case, "1 pass per a rule" might be faster... Maybe we can make smarter the overlapping detection, e.g. locked range does not prevent the same rule.

**Comment by alangpierce** [comment]
For `no-else-return` and `no-useless-return` in #8067, my reasoning was "I am modifying the control flow of this function, so I don't want any other fixes touching the function at the same time, but any changes that are outside this function should be safe.". I guess technically it's possible for a rule to look at the control flow details of a function and use that to make a modification outside of it, but that seems hard to imagine (and ideally that fixer would reserve that function's code range with the same mechanism). My gut feeling is that the `retainEnclosingFunction` method from my PR would be generally useful (and make it easy for the rule author) since plenty of changes seem isolated to their function, but I guess it's hard to say for sure.

I do really like the mindset of "what parts of the AST did I read to determine that this fix is safe". But I worry that specifying that precisely would make the rules more complex, and that the perf gain would be relatively small compared to simpler options like "lock the surrounding tokens", "lock the current function", "lock the current class", etc (which is what I had in mind with `FixTracker` in my PR, although `FixTracker` does allow more granular locking if you want).

To explain a little more where I'm coming from, I've been working on the [decaffeinate](https://github.com/decaffeinate/decaffeinate) project for a while now, which converts CoffeeScript into nice-looking modern JavaScript as a migration tool for projects that want to move off of CoffeeScript. Rather than exposing a hundred style options, decaffeinate intentionally has no style configuration, and wants you to run `eslint --fix` at the end to make the code match your preferred style. I'm trying to get decaffeinate stable enough that you can run it on hundreds of thousands of lines of code and be fairly confident that it hasn't introduced any bugs, so that means `eslint --fix` needs to be that stable as well. So far I've found three eslint autofixing bugs as part of this work (#7811, #8026, #8116).

> Just out of curiosity, how often do these fix conflicts occur when the actual fixes aren't right next to each other?

I'm working on making decaffeinate work on various open source projects (the [status page on the decaffeinate README](https://github.com/decaffeinate/decaffeinate#status) has a list), so hopefully if there are other autofix bugs like this, I'll be able to find them over the next few weeks. Currently I'm just using eslint-config-airbnb, but I could also run it with other common eslint configurations or a configuration designed to exercise as many tricky cases as possible.

**Comment by alangpierce** [comment]
To summarize the current state of PRs for this, there are two existing PRs that take different approaches to address this issue:

* #8067 is a PR from me that adds an internal builder-style class called `FixTracker`. For example, `no-useless-return` removes a `return` while stopping other updates to its function using `new FixTracker(fixer, context.getSourceCode()).retainEnclosingFunction(node).remove(node)`. The vision is that `FixTracker` would later be extended to allow multiple fix operations, then moved to the public API when we're confident with it.

* #8101 is a PR from @mysticatea that extends the autofixing API so you can optionally return an array (or iterable) of non-overlapping fixes. It also adds a `fixer.keep` method to generate a fix that doesn't change a node/token. For example, the `semi` conflict is avoided using `yield fixer.remove(lastToken)` and then `yield fixer.keep(nextToken)`. It's an API extension, although it could also be modified to not make any API changes, I think (which would give us more flexibility to iterate on it if we wanted to).

Both changes fix this issue (#7928). Mine also fixes the `no-else-return`/`no-useless-return`conflict (#8026) but doesn't add support for multiple fixes (yet). @mysticatea's PR adds support for multiple fixes (#7348) but doesn't fix the `no-else-return`/`no-useless-return` conflict (yet). Both PRs could have follow-up changes to fix all of these issues, I think, but since they take different approaches, we probably shouldn't merge both (although technically we could).

(I did have concerns about how #8101 would later address the `no-else-return`/`no-useless-return` conflict, which I left in a review there, but I think it should be possible to make something work.)

While I do still like my `FixTracker` class from #8067 and the idea of iterating on an internal API before making any public API changes, I'm not particularly tied to it, and happy to close it if #8101 seems like a better long-term option. In the near term, I'm most interested in releasing some fix for the `no-else-return`/`no-useless-return` conflict, since it's causing a test failure for me.

**Comment by mysticatea** [comment]
Actually, the main purpose of #8101 is to handle autofix as not removing comments easily (I wanted to use it in #8099). I'm not opposing to merge #7928.

**Comment by not-an-aardvark** [comment]
The bug that was initially reported in this issue was fixed by https://github.com/eslint/eslint/commit/4d35a81c9ddae2e77409bb6d82adeb62e1e1c33c, but I'm going to reopen this since I think we do need to eventually figure out a better general solution.

**Comment by alangpierce** [comment]
FWIW, since https://github.com/eslint/eslint/commit/4d35a81c9ddae2e77409bb6d82adeb62e1e1c33c was released, I've run `eslint --fix` on about 300k additional lines of code (the atom codebase run through decaffeinate and then autofixed under standardjs, and the codecombat codebase run through decaffeinate and then autofixed under eslint-config-airbnb). All tests now pass in both projects, and I haven't found any additional correctness issues with eslint autofixing. Obviously my testing isn't exhaustive, but it certainly suggests that autofix conflicts (and autofix bugs in general) are reasonably rare, at least with the rules enabled by those two configurations.

**Comment by not-an-aardvark** [comment]
Closing this because this seems to be a very rare issue with workarounds, so a more general solution isn't a priority right now.
